dialects/oracle/oracle.go

```diff
@@ -0,0 +1,223 @@
+
+	if err := s.db.QueryRow(`SELECT count(*) FROM USER_CONSTRAINTS WHERE CONSTRAINT_NAME = :1 AND constraint_type = 'R' AND table_name = :2`, foreignKeyName, tableName).Scan(&count); err == nil {
+	   return count > 0
+   } 
+   return false
+}
+
+func (s oracle) HasIndex(tableName string, indexName string) bool {
+	var count int
+	tableName = strings.ToUpper(tableName)
+	indexName = strings.ToUpper(indexName)
+	if err := s.db.QueryRow("SELECT count(*) FROM ALL_INDEXES WHERE INDEX_NAME = :1 AND TABLE_NAME = :2", indexName, tableName).Scan(&count); err == nil {
+		return count > 0
+	}
+	return false
+}
+
+func (s oracle) HasTable(tableName string) bool {
+	var count int
+	_, tableName = currentDatabaseAndTable(&s, tableName)
+	tableName = strings.ToUpper(tableName)
+	if err := s.db.QueryRow("select count(*) from user_tables where table_name = :1", tableName).Scan(&count); err == nil {
+		return count > 0
+	} 
+	return false
+}
+
+func (oracle) LastInsertIDReturningSuffix(tableName, columnName string) string {
+	return ""
+}
+
+func (oracle) LastInsertIDOutputInterstitial(tableName, columnName string, columns []string) string {
+	return ""
+}
+
+func (s oracle) ModifyColumn(tableName string, columnName string, typ string) error {
+	_, err := s.db.Exec(fmt.Sprintf("ALTER TABLE %v MODIFY %v %v", tableName, columnName, typ))
+	return err
+}
+
+func (s oracle) RemoveIndex(tableName string, indexName string) error {
+	_, err := s.db.Exec(fmt.Sprintf("DROP INDEX %v", indexName))
+	return err
+}
+
+func (oracle) SelectFromDummyTable() string {
+	return "FROM DUAL"
+}
+
+func (s *oracle) SetDB(db gorm.SQLCommon) {
+	s.db = db
+}
+
+func currentDatabaseAndTable(dialect gorm.Dialect, tableName string) (string, string) {
+	if strings.Contains(tableName, ".") {
+		splitStrings := strings.SplitN(tableName, ".", 2)
+		return splitStrings[0], splitStrings[1]
+	}
+	return dialect.CurrentDatabase(), tableName
+}
+
+func (s *oracle) DataTypeOf(field *gorm.StructField) string {
+	if _, found := field.TagSettingsGet("RESTRICT"); found {
+		field.TagSettingsDelete("RESTRICT")
+	}
+	var dataValue, sqlType, size, additionalType = gorm.ParseFieldStructForDialect(field, s)
+
+	if sqlType == "" {
+		switch dataValue.Kind() {
+		case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Uint, reflect.Uint8,
+			reflect.Uint16, reflect.Uintptr, reflect.Int64, reflect.Uint32, reflect.Uint64,
+			reflect.Float32, reflect.Float64:
+			if s.fieldCanAutoIncrement(field) {
+				sqlType = "NUMBER GENERATED BY DEFAULT AS IDENTITY"
+			} else {
+				sqlType = "NUMBER"
+			}
+		case reflect.String:
+			if _, ok := field.TagSettingsGet("SIZE"); !ok {
+				size = 0 // if SIZE haven't been set, use `text` as the default type, as there are no performance different
+			}
+			switch {
+			case size > 0 && size < 4000:
+				sqlType = fmt.Sprintf("VARCHAR2(%d)", size)
+			case size == 0:
+				sqlType = "VARCHAR2 (4000)"

```

Comment:
It would be helpful to add a comment on why 4000 is used.

Reply:
fixed.